src/main/java/org/tailormap/api/controller/PageController.java

```diff
@@ -64,51 +68,59 @@ public PageResponse page(@PathVariable(required = false) String name) {
     return pageRepository.findByName(name).map(this::getPageResponse).orElseThrow(this::notFound);
   }
 
+  private static class ViewerPageTileResult {
+    ViewerPageTile viewerPageTile;
+    boolean shouldBeFiltered;
+  }
+
   private PageResponse getPageResponse(Page page) {
     PageResponse pageResponse = new PageResponse();
     BeanUtils.copyProperties(page, pageResponse);
     pageResponse.tiles(
-        page.getTiles().stream().map(this::convert).filter(Objects::nonNull).toList());
+        page.getTiles().stream()
+            .map(this::convert)
+            .filter(viewerPageTileResult -> !viewerPageTileResult.shouldBeFiltered)
+            .map(viewerPageTileResult -> viewerPageTileResult.viewerPageTile)
+            .toList());
     return pageResponse;
   }
 
   /**
    * @param tile The page tile configuration
    * @return A page tile for the viewer, or null if the tile should not be shown (filtered).
    */
-  private ViewerPageTile convert(PageTile tile) {
+  private ViewerPageTileResult convert(PageTile tile) {
     ViewerPageTile viewerPageTile = new ViewerPageTile();
+    ViewerPageTileResult result = new ViewerPageTileResult();
+    result.viewerPageTile = viewerPageTile;
+    result.shouldBeFiltered = false;
     BeanUtils.copyProperties(tile, viewerPageTile);
 
-    if (tile.getApplicationId() != null) {
-      return Optional.ofNullable(tile.getApplicationId())
-          .flatMap(applicationRepository::findById)
-          .filter(
-              application ->
-                  !Boolean.TRUE.equals(tile.getFilterRequireAuthorization())
-                      || authorizationService.mayUserRead(application))
-          .map(
-              application -> {
-                viewerPageTile.applicationUrl("/app/" + application.getName());
-                viewerPageTile.setApplicationRequiresLogin(
-                    !authorizationService.mayUserRead(application));
-                return viewerPageTile;
-              })
-          .orElse(null);
-    }
+    Optional.ofNullable(tile.getApplicationId())
+        .flatMap(applicationRepository::findById)
+        .filter(
+            application ->
+                !Boolean.TRUE.equals(tile.getFilterRequireAuthorization())
+                    || authorizationService.mayUserRead(application))
+        .ifPresentOrElse(
+            application -> {
+              viewerPageTile.applicationUrl("/app/" + application.getName());
+              viewerPageTile.setApplicationRequiresLogin(
+                  !authorizationService.mayUserRead(application));
+            },
+            () -> result.shouldBeFiltered = true);
```

Comment:
If `applicationId` is null this will always filter the tile, even when `pageId` just `url` is set
```suggestion
        .ifPresent(
            application -> {
              boolean userNotAuthorizedForApplication =
                  !authorizationService.mayUserRead(application);
              result.shouldBeFiltered =
                  Boolean.TRUE.equals(tile.getFilterRequireAuthorization())
                      && userNotAuthorizedForApplication;
              if (!result.shouldBeFiltered) {
                viewerPageTile.applicationUrl("/app/" + application.getName());
                viewerPageTile.setApplicationRequiresLogin(userNotAuthorizedForApplication);
              }
            });
```

Reply:
This way a tile won't get filtered when the applicationId is null, but also won't get filtered (shown) when applicationId is invalid. We may put this code in just a 
```java
if(tile.getApplicationId() != null) {
  ...
}
```
For instance, a tile has a description only for authorized users. But the app (with configured authorizations) is deleted. The tile with the 'secret' description would still be shown with this suggestion...

Reply:
Fixed by surrounding with `if`